ceph design: OSD redesign and deprecating old OSD types (#3247)
rook/rook: design/ceph-deprecate-old-osd-types.md

effectively in sustaining mode, and Bluestore is the defacto replacement for it.

All of the various OSD support patterns/code-paths need changes to allow the [planned updates to
Rook's configuration strategy](/design/ceph-config-updates.md) to proceed. The various code paths
are intertangled and difficult to separate, and this presents a huge hurdle for the config changes.
In general, the intertangled OSD code poses a risk for long-term maintenance, as it may be more
prone to bugs being introduced, and it has and will continue to slow development.

In light of this background information, this design proposes three high-level goals:
1. Identify OSD features/code-paths that should be deprecated due to having become outmoded in some form
2. Propose a new architecture for Rook's OSD code that is compatible with the planned config updates
and that is extensible for the future.
3. Provide a timeline for deprecating outmoded, and provide users tools for transitioning their
   deployments which use old methods to begin using the new methods.

## Deprecated features and code paths
This design proposes to deprecate ...

> - the use of directories as storage targets

Reviewer: @travisn and I had a conversation a few weeks back about still including support for a directory for bluestore, where we just create a file for the ``block`` file/device.  This is reasonably simple and (from a ceph perspective) will be suppoortable going forward.  It does mean one non-ceph-volume OSD path, though.

Author: Yeah. Travis mentioned that this morning, and I haven't considered how to include it in this design. We did talk about effectively making that a "hidden" option that isn't exposed to users.

Reviewer 2: What if we had a simple CRD property that indicates how big of an OSD to create under dataDirHostPath? I don't think we can make it hidden, but it needs to be obviously a non-production setting.
```
  dataDirHostPath: /var/lib/rook
  dataDirHostPathStorageForTestEnv: 5Gi
```

Or perhaps that property is just a bool, and we could automatically choose a size from 1-10Gi depending on available free space, but the size matters since we have to reserve it up front.

Author: Wouldn't it still be possible to add an env var that the operator could read to the same effect? `TESTING_STORAGE_FILE_SIZE_GiB`. We could assume not present or less than one means do-not-create.

This would work fine for simple dev and test, but we would be failing to exercise a lot of OSD code if we don't test in environments with actual disks. We could also look into creating virtual disks, but I vaguely recall wanting to do that in the past and finding that it was not a straightforward process.

- support for the Filestore backing store
- support for devices not provisioned with `ceph-volume`

## New OSD operator architecture
The new OSD operator design focuses only on devices provisioned with Bluestore using `ceph-volume`.

> Some design thought is given to allowing space for support of post-Bluestore backing stores.

Reviewer 2: Another consideration is for creating OSDs on top of PVCs as described in #919, which is planned for 1.1. I'm not sure it will affect this design change, but just something to be aware of.

Author: I will have to consider this more. It's clearly a feature that users want, so I think it should go in, but I have this sinking feeling that users are going to create some very degenerate Ceph clusters with this feature, and we are going to have to explain why we can only support that to a certain point. We're already seeing one user wanting to create a Ceph cluster using a single block device on each node and then resize the underlying block device, which is absolutely bonkers from Ceph's standpoint being a distributed object store designed for real-world hardware.

With the scattered thoughts I have currently, I think the PVC-backed use case will have to be a feature of Rook-managed mode only.

Reviewer 2: With the way K8s is moving, the block PVs will help us more than just the scenarios users are asking for in #919. It helps us move into a world where K8s is aware of the underlying storage and can manage the raw devices for us. While the devil is in the details, I would expect this will also help with our permissions challenge with mounting /dev.

Author: In the update, I broke Rook-managed mode into 2 submodes: an automatic mode and a declarative mode, and the PVC use case is applicable only for the declarative mode.

mentioned in this design, a reasonable product-of-opportunity attempt may be possible, but
completion of this design is not predicated on completion of security concerns.

### Operator knowledge
Currently, the config updates are limited because the Rook operator has limited knowledge of the
hardware being configured on hosts. The updated config plan calls for a Ceph config file that is
minimal and standard for all daemons; daemons are individually configured using command line flags.
In order to properly set command line flags for object storage daemons, the operator must have
knowledge of the disks which an OSD will use. The operator must configure, for example, an OSD to
use one disk as the main storage, another disk as wal storage, and another disk as db storage.
Mechanisms are not currently in place to report the needed disk information to the operator.

### Proposed design
The existing OSD operator design uses a two-stage process. The first stage is preparation of OSDs on
a per-host basis, and the second stage is running individual OSDs. The design proposed here
separates OSD tasks into more stages.

> #### Stage 1: Host device inspection

Reviewer 2: Are you saying the osd prepare job will be broken into stages 1-3, each with their own jobs? I'm not following how that helps things and it seems more complex. If any of stages 1-3 needs the same privileges as the prepare pod has today (such as access to /dev), it doesn't seem like we're saving anything for future security concerns.

The philosophy around the current prepare pod is:
- The operator passes the job the desired state from the CR
- The prepare pod looks at all the devices, determines what should be provisioned according to the desired state, and then goes ahead and provisions them.
- The list of OSDs created is saved to a configmap
- The operator then reads the configmap and creates the individual OSD deployments.

There will of course be  some refactoring and code cleanup inside the prepare pod as we deprecate the old way of doing things, but I don't understand the need for splitting the job.

Author: The specific security thoughts I had in mind were that stage 1 need only have *read-only* access to `/dev`. Stage 2 is just processing, and then stage 3 needs to have *read-write* access to *only* the devices it intends to provision and doesn't need write access to the entirety of `/dev`. Having write access to the entirety of `/dev` is the security hole I hope to fix here.

> The list of OSDs created is saved to a configmap

What configmap is that, by the way? I didn't see anything obvious in my cluster when I ran `kubectl -n rook-ceph get cm`.

binaries. Existing logic distribution has contributed to operation which is not always transparent
or easy to extend.

#### Stage 3: Host device preparation
This stage is one in which `ceph-volume` provisions the devices. As with stage 1, this design calls
for a Rook OSD preparation binary. The binary's single purpose is to call `ceph-volume` to provision
devices and store a post-preparation device report from `ceph-volume` into a ConfigMap (possibly the
same ConfigMap from step 1) for the operator to inspect later.

The binary should act as a simple pass-through for `ceph-volume` arguments to keep the binary's
functions as small as possible; the Ceph operator should contain the majority of the logic related
to Ceph's configuration. Arguments to `ceph-volume` here will be based on the host devices found in
stage 1 and processing done in stage 2.

For future security, this preparation pod should only need access to the individual disks which
`ceph-volume` will use to provision Ceph.

#### Stage 4: Start OSDs
The last step remaining is to start the OSDs previously created by `ceph-volume`. The

> post-preparation device report stored in step 3 will be used to determine which osds to start on any

Reviewer: I wonder if instead of a post-preparation report, it makes sense to skip back to step 1.  Inspection should then report prepared+ready devices that are just missing the OSD pods.  This would avoid some of the complexity around the post-preparation report and ceph-volume incantations that, say, produce 4 OSDs for a single device.

Author: That makes a lot of sense. Then we can call `ceph-volume` directly from a pod entrypoint and reduce the number of supplementary binaries Rook has. I'll include this in my next go-round.

the new OSD design is operational in Rook and is the only code path used for new OSDs, the
deprecation will be official, and that should be reported to users in the release notes and upgrade
guide. After there is documentation for manually migrating OSDs, we can expect that the following
version of Rook will officially drop support for legacy OSDs.

It is ideal if users have access to the migration utility at the time of deprecation; however, the
utility needn't be released at the same time as Rook since it is a separate utility. The utility
must be released before or at the same time as the Rook release which officially drops support for
legacy OSDs. The Rook release which officially drops this support should perform a preflight check
to ensure that no legacy OSDs exist, and if it finds existing legacy OSDs, it should error out with
a message instructing users to migrate.

### OSD migration

<!-- HALP! I need some help here figuring out how disks will be removed in practice -->

For nodes with only directory-based OSDs, the user will need to attach some number of new disks to
those nodes and wait for Rook to provision the new disks for use. The directory-based OSDs will be

> migrated by out-ing one directory-based OSD at a time from Ceph and waiting for Ceph to report that

Reviewer: nit: if it is known that the OSD isn't coming back, then instead of ``ceph osd out osd.NNN`` the user/operator should do ``ceph osd crush rm osd.NNN``.

Author: I'm not super familiar with OSD operations, so that's a critical detail for me. Thanks. :)

respectively. In the **Rook-managed** pattern, Rook will continue to offer simplified management
operations in the same manner as Rook does today, abstracting some of the smaller details from the
user. In the **Ceph-managed** pattern, crucial for users with strict performance requirements of any
kind, the user will use Ceph's CLI and/or dashboard to manage Ceph daemons and data pools
explicitly. This will look more familiar to the way existing Ceph clusters are managed today.

##### Rook-managed mode
This mode is characterized by either of the following Ceph cluster configuration options:
`useAllDevices: true` or `deviceFilter: <set>`. This means that every applicable unused device is
turned into an OSD, on whichever nodes are appropriate (either specifically-labeled or `useAllHosts:
true`).  In addition, any device that is found by the agent on any host that is labeled with the
Ceph cluster UUID and corresponds to a valid OSD in the cluster's OSDMap should be started.

<!-- discuss -->
Rook may opt to delete OSDs which have been `out` for longer than some threshold period measured in
days, e.g., 45 days.

> ##### Ceph-managed mode

Author: I think it partially handles this currently. There is still the need for Rook to be able to create/delete/replace OSDs upon request by the Ceph mgr.

> ##### Ceph-managed mode

Author: I wonder as well, but I haven't heard or come up with anything that seems like it makes sense. What do you think about my latest overall comment about shifting our mindset to thinking of the disk itself as the declaration?

> ##### Ceph-managed mode

Reviewer 2: > Rook users may still wish to add disks to their Ceph clusters as
OSDs, but this design proposes to remove the deletion-by-omission functionality. Disks which users
select for their Ceph clusters can be viewed as a declaration of intent to **add** disks as OSDs,
but once disks are added, the disk content itself becomes the declaration of **existence** of OSDs.

Removing OSDs automatically has never felt quite right to me. Turning off this behavior sounds right. As you said, then the CR declaration is all about **adding** OSDs. 
- If you want to replace an OSD, remove the old disk and plug in your new disk. The prepare job could recognize the swap and call ceph-volume with the osd id to replace (details still need to be worked out to automate this, but this is the desired behavior)
- If you plugin a new disk that matches the osd filter or update the osd filter, an orchestration would add those osds
- If you remove an OSD that you won't be replacing, rook would do nothing with the crush map. Rook wouldn't only clean up the k8s deployment if the osd is safe-to-destroy (see #3117) 

Are there other implications of your proposal?

Atop all of this, there have been discussions ongoing about supporting Ceph's Rook
orchestrator as the intended way of managing Ceph deployments on Kubernetes. Orchestration currently
modifies the Ceph cluster resource, but from the standpoint of a Ceph storage administrator, the CRD
(and Rook) lacks the flexibility needed for non-trivial Ceph deployments. While flexibility could be
added, it will be better to remove the need for such complexity in the cluster CRD and in Rook. The
details of this part of the design will be expounded later.

In light of this background information, this design proposes three high-level goals:
1. Identify OSD features/code-paths that should be deprecated due to having become outmoded in some
   form.
2. Propose a new architecture for Rook's OSD code that is compatible with the planned config updates
   and that is extensible for the future.
3. Provide a timeline for deprecating outmoded, and provide users tools for transitioning their
   deployments which use old methods to begin using the new methods.

## Deprecated features and code paths
This design proposes to deprecate ...

> - the use of directories as storage targets

Reviewer 3: Hm, that's not very good for small setups, where you don't have separated drives to format. Use of directories as storage targets was main argument for me for choosing Rook over OpenEBS. Using directories also helps utilizing OS partition storage, which with some setups might be significant (think bare metal server with 2TB in RAID 1).

Author: This proposal introduces a replacement for this idea of installing bluestore devices in files. Take a look and let us know if you see that as being feasible.

The reasoning behind this is that supporting directories as storage targets using Ceph is a kind of degraded use-case that to my knowledge isn't supported by Ceph. My understanding is that the feature exists in Ceph primarily for test cases. Even installing bluestore in a file isn't really "supported" but exists for testing scenarios. My concern is that Rook could be relying on unsupported behavior in Ceph for a supported (or seemingly supported) use-case, which means we may be putting user data at unnecessary risk without having users able to make informed decisions for their data. I also believe that using directories as OSDs risks easily running out of space on the OS disk. Relatedly, Ceph's usage reporting probably isn't accurate in this case since Ceph isn't designed for this use-case.

Something we could also discuss here is talking about support for using partitions as osds, which would be a safer solution for user data and OS data. I'm not sure if there are Ceph concerns related to partition use.

Reviewer 4: Yeah, `directories:` are still needed for the future at least for testing purposes only as a good amount of people are using it.

Author: Again, this proposal does have the test use case in mind. For testing purposes, we are proposing using a preallocated file as the bluestore target.

Reviewer 3: Thank you for your explanation @BlaineEXE.

While preallocated file option is not ideal, as you lose the flexibility of storage (you need to decide how much space you dedicate for ceph upfront), it's better than requiring store device (maybe loop can be used to skip preallocated file?)

I understand that this decision is driven by upstream Ceph and the proposal absolutely make sense in this context.

Author: Could a production-supportable solution be to allow partitions as storage targets? Can `ceph-volume` use partitions? @leseb @jan--f

Reviewer 5: @BlaineEXE `ceph-volume` can essentially create OSDs on any LV. If its acceptable for smaller deployments or testing to create the wanted LVs manually, the underlying storage is up to the user. This might however make the specification of which storage to use more complex.

- support for the Filestore backing store

> - support for devices not provisioned with `ceph-volume`

Reviewer 6: I'm not sure we need to deprecate old-fashioned prepared OSDs. Just like the transition between `ceph-disk` and `ceph-volume`, OSDs continue to operate. So we should be able to *detect* and *activate* these OSDs but not allow the preparation (already done since `ceph-volume` is the default.

Author: As clarity, this design proposes to deprecate these but not after providing users a means of migrating from old to new.

## New OSD operator architecture
The new OSD operator design focuses only on devices provisioned with Bluestore using `ceph-volume`.

> Some design thought is given to allowing space for support of post-Bluestore backing stores.

Reviewer 6: I don't think we should bother much, Rook will simply go with whatever `ceph-volume` default backing store will be in the future. Basically, the backing store should not be configurable in Rook.

Author: I would be in favor of this, but I'd like to get @travisn 's opinion on this topic. This could result in clusters running mixed bluestore/<nextstore> devices in the future, but this would be a far future I believe. Rook users who wish to use Rook-managed modes would gain some simplicity in that they no longer have to learn what bluestore is and no longer have to consider that as a configuration option.

explicitly. This will look more familiar to the way existing Ceph clusters are managed today.

##### Rook-managed mode
Rook-managed mode is further broken down into two sub-modes, **Automatic Rook-managed mode** and
**Declarative Rook-managed mode**.

###### Automatic Rook-managed mode
This mode is characterized by either of the following Ceph cluster configuration options:
`useAllDevices: true`, or`deviceFilter: <set>`. This means that every applicable
unused device is turned into an OSD, on whichever nodes are appropriate (either specifically-labeled
or `useAllHosts: true`). In addition, on any host that is available to the Rook cluster, any device
labeled with the Ceph cluster UUID and corresponds to a valid OSD in the cluster's OSDMap should be
started.

Rook may opt to delete OSDs which have been `out` for longer than some threshold period measured in
days, e.g., 45 days. Rook will occasionally report in its logs (at least once per day) that it intends to delete an OSD
after some number of days.

###### Declarative Rook-managed mode
This mode is characterized by the following Ceph cluster configuration options:

> `useAllDevices: false`, `deviceFilter: <unset>`, and `devices: <set>`. This means that only devices

Reviewer 6: To me `deviceFilter` and `devices` could be merged into a single entity. I know it's a matter of variable *type* but it'd nice.
Also, I believe `deviceFilter` is a form of declarative.

Author: I disagree here that `deviceFilter` is declarative. At least what I mean with declarative here is that each disk is declared, and disks could potentially be removed via omission. I suppose the same is possibly true with `deviceFilter`, but I can't imagine anyone wanting to use it that way in reality. To me `deviceFilter` is an extension of the idea "use all of the disks available" the same as `useAllDevices`, but focused on eliminating unwanted disks, resulting in a subset, "use all disks found in this subset." I could foresee a common device filter being `/dev/sd[c-z][a-z]*` as an example, and that is much closer to `useAllDevices` than to specifying devices individually.

The most critical design choice which can help reduce pod permissions is to give pods access to only
the host storage that is required for a given operation. This may mean mounting only certain paths
and/or devices into pods, and it may mean mounting paths/devices in read-only mode when possible.

Reasonable attempts to secure Rook during execution of this design can be made, but it is not in the
scope of this design that many hours be spent focusing on security initially. Where security is
mentioned in this design, a reasonable product-of-opportunity attempt may be possible, but
completion of this design is not predicated on completion of security concerns.

#### Test environments without disks
Rook must occasionally support test environments without disks. Users new to Rook especially may not
want to create block devices for their initial test environments. Ceph Bluestore has the ability to
use files as disks for such test environments, and this design proposes to enable this as an option
for such environments.

To enable this feature, the Rook-Ceph operator will observe the environment variable

> `TEST_ENVIRONMENT_FILE_DEVICE_SIZE_GiB`. If this variable is set with a positive, nonzero value, the

Reviewer 6: Bluestore block size default is 10GB, this seems reasonable for developers so I would not make this configurable. I'd like to highlight one of your earlier comments @BlaineEXE, the target audience for this must be developers only, no one should be encouraged to use this in production, keeping the default 10GB block size is one way.

Thus I'd turn this variable to something like: `TEST_ENVIRONMENT_OSD_ON_FILE`.

Author: Per my discussion with @invidian, some users may still wish to use this feature occasionally, and some flexibility to set the size would be desired. 

We could assume to round up to the nearest 10GB within the operator to better handle the 10GB block size.

When you say 10GB, do you mean Gigabytes or Gibibytes?

In order to properly set command line flags for object storage daemons, the operator must have
knowledge of the disks which an OSD will use. The operator must configure, for example, an OSD to
use one disk as the main storage, another disk as wal storage, and another disk as db storage.
Mechanisms are not currently in place to report the needed disk information to the operator.

### Proposed Rook operator design
The existing OSD operator design uses a two-stage process. The first stage is preparation of OSDs on
a per-host basis, and the second stage is running individual OSDs. The design proposed here
separates OSD tasks into more stages. The stages are numbered from the perspective of the
**Rook-managed mode**, and not all stages will apply to the **Ceph-managed mode**.

A Rook OSD prepare binary currently handles the equivalent of stages 1 through 4, and in early
implementations of the new OSD design, a version of the current binary could be used for stages 1
through 3 to achieve the minimum viable product more quickly. Splitting the stages can be done
afterwards as makes sense.

#### Stage 1: Host device inspection (Rook-managed mode only)

> The first stage is an inspection stage that is currently done as part of the existing preparation

Reviewer 6: Perhaps we should mention that `useAllDevices` is set to `true` here?

stage. Because all of the host's devices must be inspected, this stage likely needs access to the
entirety of the host's `/dev` directory. It may also need access to large sections of other

> directories like `/run/udev`.

Reviewer 6: I think `/run/udev` is really a detail so we can remove it. Also, we might need more bindmounts.

Author: `/run/udev` here isn't intended to be a specific implementation detail but an acknowledgment (and example of) something else that might be required, and the implementer of the design shouldn't assume that they are explicitly limited to `/dev`.

Similar to the current Rook OSD prepare binary, this proposal calls for a Rook inspection binary

> which will examine the disk hardware, rejecting disks that are already in use by applications other

Reviewer 6: We need to highlight that as per https://github.com/rook/rook/issues/2606 the device detection will use `ceph-volume inventory` command instead of the current detection with `lsblk`.

than Rook (as per current behavior). The inspection binary stores a host device report in a

> ConfigMap which will be read by the operator. The report need merely contain a list of devices which

Reviewer 6: ```suggestion
ConfigMap which will be read by the operator. The report merely contains a list of devices which
```

are able to be provisioned for Ceph's use (e.g., `/dev/sdc` or `/dev/disk/by-id/blah`).

Because `ceph-volume` is ultimately responsible for preparing the devices, it is ideal if
`ceph-volume` does this inspection itself and produces a report detailing the disks it could use
for later provisioning. The inspection binary should store `ceph-volume`'s report into the
aforementioned ConfigMap.

> As a matter of future security, the pod responsible for inspection should have read-only access to

Reviewer 6: It's really TBD and highly depends on `ceph-volume` availability to do this.

`/dev` (and `/run/udev` if needed). This stage should not have write access to any part of the host
unless it is absolutely required.

#### Stage 2: Process device inspection report (Rook-managed mode only)

> In the second stage, the operator examine the host device inspection report in the ConfigMap. The

Reviewer 6: ```suggestion
In the second stage, the operator examines the host device inspection report in the ConfigMap. The
```

primary inspection processing done would be to filter the list of devices reported based on the
`deviceFilter` cluster storage setting. This stage can also compare the host `devices` specified in
the cluster config to the devices found during inspection.

The insertion of stage 2 executed within the operator itself, placed in between stages 1 and 3
allows for easy security improvements later, and it also helps keep the majority of the Ceph
configuration logic in the operator itself rather than distributing the logic between multiple
binaries. Existing logic distribution has contributed to operation which is not always transparent
or easy to extend.

#### Stage 3: Host device preparation (Rook-managed mode only)
This stage is one in which `ceph-volume` provisions the devices. Unless a custom Rook binary is
absolutely necessary, `ceph-volume` can be called directly from the pod's container entrypoint.
Arguments to `ceph-volume` will be based on the host devices found in stage 1 and processing done
in stage 2.

> For future security, this preparation pod should only need access to the individual disks which

Reviewer 6: I don't think this is possible since `ceph-volume` consumes a raw block and spits out an LV, I'm not 100% sure but I believe the preparation step requires access to that LV.

Repeat stage 1 inspection. For the **Ceph-managed mode**, this will be the first and only
inspection stage. The `ceph-volume` report stored in the ConfigMap may now report existing
OSDs on the host. These OSDs may have been created by Rook in Rook-managed mode, or they may have
been created manually or via an action of Ceph's Rook orchestrator. The device report should contain
the information necessary to determine which flags should be passed to `ceph-osd`.

#### Stage 5: Start OSDs
The last step remaining is to start the OSDs which were found by `ceph-volume`'s prior inspection.
The device inspection report stored in step 4 will be used to determine which OSDs on the host
belong to the operator's Ceph cluster (by UUID examination), and appropriate deployments will be
created for each OSD.

If possible, the existing trend of using init containers instead of a separate Rook binary
(trend established for Ceph mons, mgrs, MDSes, and RGWs) should be used here with the intent being
again to keep most of the logic in the operator. If teardown operations are needed, the Kubernetes
container `PreStop` lifecycle hook should be used if reasonably possible.

> ### Proposed Ceph mgr Rook orchestrator design

Reviewer 6: Isn't this out of scope?

Author: I think this is really the core thing we are trying to achieve by v1.2 to integration with Octopus' operator. I think it's the complication of implementing this feature that results in the rest of the doc . The existing OSD code is hindering reasonable progress here, so a complete redesign is proposed.

Author: I've been discussing orchestrator things with Sebastian Wagner and Ricardo Dias this week, and I think the answer to your question is both yes and no. Parts of what I have written here should be part of an orchestrator design doc, but I think we do need to have some plans in place for how an OSD redesign in Rook can support the known orchestrator planning. I'm planning to write up an introductory orchestrator design to send to Sebastian and Ricardo and let them fill in more technical details, and I will have to modify this to be focused only on how Rook can support requests from the orchestrator.

Ceph's Rook orchestrator will require means of listing and preparing host devices for use by the
Ceph cluster. This design proposes to invoke individual runs of **stage 1** and **stage 3** upon
user input via Ceph CLI or Ceph dashboard orchestration events. Ceph's Rook orchestrator will also
require means of deleting OSDs.

This design proposes to revive the general concept of a Rook API for these invocations. By using the
already-running Rook operator to run stages it is already capable of running, the Ceph mgr pod does
not need to obtain privileges needed to create Kubernetes deployments/jobs itself.

In the design outlined below, the Ceph mgr will need the ability to create and update an
"orchestrator-osd-requests" ConfigMap which contains a queued list of actions it wishes the Rook
operator to take. In a special control loop active only in **Ceph-managed mode**, Rook will watch
for changes to the queue and take appropriate actions.

#### Adding OSDs
Ceph will update the "orchestrator-osd-requests" ConfigMap noting the OSD(s) it wishes the Rook

> operator to create on a given host and the `ceph-volume` commands that should be run to create them.

Reviewer 6: ```suggestion
operator to create on a given host and the `ceph-volume` command that should be run to create them.
```

update, at which time the Ceph orchestrator can now read the most up-to-date inspection report valid
for all nodes. Updating the device list takes priority over creation, deletion, and replacement.

> **TODO**
> how will we handle atomicity of configmap updates?
>
> one option would be to have the mgr be the only entity allowed to edit the configmap. Queued items
could be removed once Ceph is aware a device has been added, deleted, or replaced. The mgr can be
aware of updates to node device lists when the timestamps for all nodes' reports are after the
timestamp for the request to list devices. In this case, Rook would have to keep internal track of
which queue items it had already completed so that it could continue working on queue items while
the Ceph mgr has not yet removed items from the queue.
>
> Another option would be to handle adding/removing queue items only by using patches to the
configmap. I'm not certain this is possible, but it could be investigated.

## Deprecation plan
The first step in the deprecation plan is to have the new design in-place. Once the new design is
available, the new design should always be used for any new OSD devices. The overall deprecation

> plan calls for pre-existing user OSDs to be migrated from the newly deprecated OSDs to OSDs using

Reviewer 6: Are you saying people to destroy old-fashion prepared OSD and re-create them with the new design? I wouldn't do this.

Author: That is the intent here.

Author: For some clarity, this would be filestore and directory-based OSDs only.

the new design.

> Regardless of progress on completing this design, users should be informed in the release notes and

Reviewer 6: Again, I don't think this is possible nor reasonable to ask. Unfortunately, this is our burden/legacy and we have to deal with that. Users shouldn't suffer from bad decisions from the past :-/.

Author: We could still allow these old OSDs possibly. My question is whether remove operations would be different for legacy disks versus "new" disks. I suppose in the end Rook just needs to remove the k8s resources and the orchestrator will handle the OSD removal from CRUSH or the osd `out`-ing.
